src/relay/pass/partial_eval.cc

```diff
@@ -0,0 +1,714 @@
+  }
+
+  PStatic VisitExpr_(const LetNode* op, LetList* ll) final {
+    env_.Insert(op->var, VisitExpr(op->value, ll));
+    return VisitExpr(op->body, ll);
+  }
+
+  PStatic VisitExpr_(const IfNode* op, LetList* ll) final {
+    PStatic c = VisitExpr(op->cond, ll);
+    if (c->pstatic) {
+      DLContext cpu_ctx;
+      cpu_ctx.device_type = kDLCPU;
+      cpu_ctx.device_id = 0;
+      NDArray cpu_array = c->pstatic->get<STensorNode>().data.CopyTo(cpu_ctx);
+      CHECK_EQ(TVMType2Type(cpu_array->dtype), Bool());
+      // TODO(@jroesch, @MK): Refactor code into helper from DCE.
+      if (reinterpret_cast<uint8_t*>(cpu_array->data)[0]) {
+        return VisitExpr(op->true_branch, ll);
+      } else {
+        return VisitExpr(op->false_branch, ll);
+      }
+    } else {
+      Expr t = store_.Extend<Expr>([&](){ return VisitExpr(op->true_branch, ll)->dynamic; });
+      Expr f = store_.Extend<Expr>([&](){ return VisitExpr(op->false_branch, ll)->dynamic; });
+      store_.Invalidate();
+      return NoStatic(ll->Push(IfNode::make(c->dynamic, t, f)));
+    }
+  }
+
+  PStatic VisitExpr_(const RefCreateNode* op, LetList* ll) final {
+    PStatic ps = VisitExpr(op->value, ll);
+    Static r = SRef();
+    store_.Insert(&r->get<SRefNode>(), ps);
+    return HasStatic(r, RefCreateNode::make(ps->dynamic));
+  }
+
+  PStatic VisitExpr_(const RefWriteNode* op, LetList* ll) final {
+    PStatic r = VisitExpr(op->ref, ll);
+    PStatic v = VisitExpr(op->value, ll);
+    if (r->pstatic) {
+      store_.Insert(&r->pstatic->get<SRefNode>(), v);
+    } else {
+      store_.Invalidate();
+    }
+    return HasStatic(STuple({}), RefWriteNode::make(r->dynamic, v->dynamic));
+  }
+
+  PStatic VisitExpr_(const RefReadNode* op, LetList* ll) final {
+    PStatic r = VisitExpr(op->ref, ll);
+    if (r->pstatic) {
+      PStatic ret = store_.Lookup(&r->pstatic->get<SRefNode>());
+      if (ret) {
+        return ret;
+      }
+    }
+    return NoStatic(RefReadNode::make(r->dynamic));
+  }
+
+  PStatic VisitExpr_(const CallNode* op, LetList* ll) final {
+    PStatic f = VisitExpr(op->op, ll);
+    std::vector<PStatic> x;
+    tvm::Array<Expr> x_dyn;
+    for (const Expr& e : op->args) {
+      PStatic ps = VisitExpr(e, ll);
+      x.push_back(ps);
+      x_dyn.push_back(ps->dynamic);
+    }
+    if (f->pstatic) {
+      return f->pstatic->get<SClosNode>().func(x, op->attrs, op->type_args, ll);
+    } else {
+      store_.Invalidate();
+      return NoStatic(CallNode::make(f->dynamic, x_dyn, op->attrs, op->type_args));
+    }
+  }
+
+  PStatic VisitExpr_(const FunctionNode* op, LetList* ll) final {
+    if (op->IsPrimitive()) {
+      return HasStatic(SClos(ConstEvaluateClos(GetRef<Expr>(op), ll)), GetRef<Expr>(op));
+    } else {
+      Clos f = [=](const std::vector<PStatic>& pv,
+                   const Attrs& attrs,
+                   const tvm::Array<Type>& type_args,
+                   LetList* ll) {
+        return env_.Extend<PStatic>([&]() {
```

Comment:
I could be mistaken, but it looks like all of the values in scope are references. It'll be more efficient to just copy them using `[=]` instead of taking `&&` with `[&]` (but my guess is that it doesn't actually matter because compilers are good).

Reply:
I'd like & as it is more 'extensible': if you add a value outside that got used it wont copy it